Add initial documentation scaffolding (#18)
bulk-dqm/bulk-data-export-client: README.md

<p align="center">
  <h1 align="center">Bulk Data Export Client</h1>
<div align="center">

CLI application for FHIR Bulk Data Export and FHIR-based quality measure calculation using the HL7® FHIR® standard<sup id="fn-1">[\[1\]](#fnref-1)</sup>

</div>

> </p>

Reviewer: I really like the centered look!

---

# Table of Contents
- [Introduction](#introduction)
- [Installation](#installation)
  - [Prerequisites](#prerequisites)
  - [Local Installation](#local-installation)
  - [Local Usage](#local-usage)
- [Running the Client via the CLI](#running-the-client-via-the-cli)
- [Configuration](#configuration)
  - [Creating a Custom Configuration File](#creating-a-custom-configuration-file)
  - [CLI Options](#cli-options)
- [Three-Step Workflow](#three-step-workflow)
- [Dependent Open-Source Libraries](#dependent-open-source-libraries)
  - [FHIR Quality Measure (FQM) Execution](#fhir-quality-measure-fqm-execution)
  - [SMART on FHIR Bulk Data Client](#smart-on-fhir-bulk-data-client)
- [API Reference](#api-reference)
- [Advanced Topics](#advanced-topics)
- [Documentation Resources](#documentation-resources)
- [License](#license)

## Introduction

> `bulk-data-export-client` is a reference implementation of a FHIR client that requests a bulk-data export from a server, receives status information regarding processing the requested files, and downloads the retrieved files. The capabilities of the client are restricted to Group-level data export.

Reviewer: ```suggestion
`bulk-data-export-client` is a prototype implementation of a FHIR client that requests a Bulk Data Export from a server, receives status information regarding processing the requested files, and downloads the retrieved files. The capabilities of the client are restricted to Group-level data export.
```

Nitpicking, but I think it would be more accurate to call this a prototype implementation rather than a reference implementation.
[See NIST's definition of a reference implementation](https://csrc.nist.gov/glossary/term/Reference_Implementation)
> the implementation of a standard to be used as a definitive interpretation for the requirements in that standard.

Also, changing instances of "bulk-data export" to "Bulk Data Export" because that's how it seems to be [consistently referred to in the spec](https://hl7.org/fhir/uv/bulkdata/export.html).

> Optionally, the client can perform calculation on FHIR-based Electronic Clinical Quality Measures (eCQMs) written in Clinical Quality Language (CQL) using the retrieved data from a successful bulk-data export operation. For more information on measure calculation, see the [fqm-execution](https://github.com/projecttacoma/fqm-execution) calculation library.

Reviewer: ```suggestion
Optionally, the client can perform calculation on FHIR-based Electronic Clinical Quality Measures (eCQMs) written in Clinical Quality Language (CQL) using the retrieved data from a successful Bulk Data Export. For more information on measure calculation, see the [fqm-execution](https://github.com/projecttacoma/fqm-execution) calculation library.
```

node build/cli.js [options]
```

## Running the Client via the CLI

> The following steps outline the end-to-end workflow for sending a bulk-data export request to a bulk-data server:

Reviewer: ```suggestion
The following steps outline the end-to-end workflow for sending a Bulk Data Export request to a bulk-data server:
```

**1. Specify configuration options**

There are two ways to specify configuration options:
- using CLI flags (if defined)

> - by specifying the configuration values in a configuration file

Reviewer: ```suggestion
- using a configuration file
```

Improving the parallelism of the bullets

To use values from a configuration file, use the `--config` CLI flag to specify the relative path of the desired configuration file to use. 

> See [configuration-options] for an overview of all supported options.

Reviewer: ```suggestion
See [Creating a Custom Configuration File](#creating-a-custom-configuration-file) for an overview of all supported options.
```

Fixing broken link

> See [cli-options] for an overview of all supported CLI options.

Reviewer: ```suggestion
See [CLI Options](#cli-options) for an overview of all supported CLI options.
```
Fixing broken link

**2. Run the CLI**

> The user can run the application end-to-end or run a subset of steps. To see which CLI options are required at each step, see [three-step-workflow]. For this example, we will assume we want to run the end-to-end workflow (i.e. run all three steps).

Reviewer: ```suggestion
The user can run the application end to end or run a subset of steps. To see which CLI options are required at each step, see [three-step-workflow]. For this example, we will assume we want to run the end-to-end workflow (i.e. run all three steps).
```

The english stackexchange tells me that[ "end-to-end" is an adjective while "end to end" is an idiom](https://english.stackexchange.com/questions/531950/end-to-end-vs-end-to-end).

This seems to check out based on the [definition of end to end](https://www.merriam-webster.com/dictionary/end%20to%20end) and [rules for compound adjectives](http://www.kentlaw.edu/academics/lrw/grinker/LwtaCompound_Adjectives.htm).

(Trying to learn and improve my own grammar while reviewing this. Past me would be shocked.)

> To run the application end-to-end, we must specify a valid FHIR url, a group id that corresponds to a FHIR Group resource on the server, and a path to a measure bundle that will be used for measure calculation.

Reviewer: ```suggestion
To run the application end to end, we must specify a valid FHIR URL, a group id that corresponds to a FHIR Group resource on the server, and a path to a measure bundle that will be used for measure calculation.
```

> Let's assume the FHIR url we want to use for export is `http://my-fhir-server`, the group we are interested in exporting data for has id `my-group`, and the path to our measure bundle is `path/to/my-measure-bundle`. One way to specify our command using the CLI is:

Reviewer: ```suggestion
Let's assume the FHIR URL we want to use for export is `http://my-fhir-server`, the group we are interested in exporting data for has id `my-group`, and the path to our measure bundle is `path/to/my-measure-bundle`. One way to specify our command using the CLI is:
```

| patientBundles | string | Directory containing patient bundles. |
| tokenUrl | string | The Bulk Data server token URL ("none" for open servers) |
| privateKey | string | The private key used to sign authentication tokens. |
| clientId | string | The client ID |
| scope | string | The scope to use in the authorization request. |
| accessTokenLifetime | number | The access token lifetime in seconds. |
| reporter | string | Reporter to use to render the output. |
| _outputFormat | string | The value of the `_outputFormat` parameter for Bulk Data kick-off requests. |
| _since | string | The value of the `_since` parameter for Bulk Data kick-off requests. |
| _type | string | The value of the `_type` parameter for Bulk Data kick-off requests. |
| _elements | string | The value of the `_elements` parameter for Bulk Data kick-off requests. |
| patient | string | The value of the `patient` parameter for Bulk Data kick-off requests. |
| includeAssociatedData | string | The value of the `includeAssociatedData` parameter for Bulk Data kick-off requests. |
| _typeFilter | string | The value of the `_typeFilter` parameter for Bulk Data kick-off requests. |

> | group | string | Id of FHIR group to export. |

Reviewer: ```suggestion
| group | string | Id of FHIR Group to export. |
```

| forceStandardFileNames | boolean | Specifies whether to rename files to match the standard naming convention `{fileNumber}.{ResourceType}.ndjson`.
| downloadAttachments | boolean | Specifies whethr to download external attachments found in `DocumentReference` resources. |
| inlineDocRefAttachmentsSmallerThan | number | In `DocumentReference` resources, any `attachment` elements having a `url` (instead of inline data) and a `size` below this number will be downloaded and put inline as base64 `data`. |
| inlineDocRefAttachmentTypes | string[] | Only files of listed types will be inlined and the rest will be downloaded into "attachment" subfolder. |
| pdfToText | boolean | Specifies whether attachments of type PDF that are being inlined will first be converted to text and then inlined as base64. |
| destination | string | Path to download destination. |
| awsRegion | string | AWS Region (only used if `destination` points to S3). |
| awsAccessKeyId | string | AWS Access Key ID (only used if `destination` points to S3). |
| awsSecretAccessKey | string | AWS Secret Access Key (only needed if `destination` points to S3). |
| logFile | string | Path to a log file to write logs to. |

> | outputPath | string | Path to an output file for generated FHIR `MeasureReport`s. |

Reviewer: ```suggestion
| outputPath | string | Path to an output file for generated FHIR MeasureReports. |
```

I think the Markdown code style makes this clunky with the "s" appended.

## Three-Step Workflow
The CLI logic can be categorized into three main processes:

> 1. Execute bulk data $export by interacting with a FHIR server. The exported NDJSON files are downloaded to the specified download directory.

Reviewer: ```suggestion
1. Execute Bulk Data $export by interacting with a FHIR server. The exported NDJSON files are downloaded to the specified download directory.
```

2. Create FHIR patient bundles from the exported NDJSON. The FHIR patient bundles are saved to the specified patient bundles directory.
3. Run measure calculation on the FHIR patient bundles using the fqm-execution measure calculation library

> The user can run the application end-to-end or run a subset of these steps. To run a subset of the steps, follow the CLI configuration options below. While it does not take a lot of time to run the application end-to-end on a small number of patients, the amount of time needed to run the application greatly increases depending on the number of patients contained in the FHIR group and the server's implementation of bulk export.

Reviewer: ```suggestion
The user can run the application end to end or run a subset of these steps. To run a subset of the steps, follow the CLI configuration options below. While it does not take a lot of time to run the application end to end on a small number of patients, the amount of time needed to run the application greatly increases depending on the number of patients contained in the FHIR group and the server's implementation of Bulk Data Export.
```

#### Running Step 3 Only
Required CLI flags:
```bash
-b, --patient-bundles <patient-bundles>
-m, --measure-bundle <measure-bundle>
```
This uses the FHIR patient bundles from the patient bundles directory along with the specified measure bundle to perform measure calculation with the [fqm-execution measure calculation library](https://github.com/projecttacoma/fqm-execution). Produces a summary measure report on all the patient bundles that are available.

#### Running Steps 1 and 2
Required CLI flags:
```bash
-f, --fhir-url <FHIR URL>
-g, --group-id <id>
-b, --patient-bundles <patient-bundles>
```
The `-m` flag should not be provided, as it triggers Step 3 to be executed.

#### Running Steps 2 and 3
Required CLI flags:
```bash
-d, --destination <destination>
-m, --measure-bundle <measure-bundle>
```

#### Running all 3 Steps
Required CLI flags:
```bash
-f, --fhir-url <FHIR URL>
-g, --group-id <id>
-m, --measure-bundle <measure-bundle>
```
The `-d` and `-b` flags may also be provided to specify storage directories. If not provided, the end-to-end workflow will use the default directory paths.

## Dependent Open-Source Libraries

### FHIR Quality Measure (FQM) Execution

> `fqm-execution` is a library for calculating eCQMs witten in CQL using the HL7® FHIR® standard. The library is used to generate a FHIR summary [Measure Report](https://www.hl7.org/fhir/measurereport.html) on the patients referenced in the FHIR group used for export.

Reviewer: ```suggestion
`fqm-execution` is a library for calculating eCQMs witten in CQL using the HL7® FHIR® standard. The library is used to generate a FHIR summary [MeasureReport](https://www.hl7.org/fhir/measurereport.html) on the patients referenced in the FHIR Group used for export.
```

`fqm-execution` is available on [npm](https://www.npmjs.com/package/fqm-execution) and [GitHub](https://github.com/projecttacoma/fqm-execution).

### SMART on FHIR Bulk Data Client

> SMART on FHIR is an open-source standards-based API. Their bulk-data client is used as a backbone for this client. Imported logic includes kick-off execution, download logic, and CLI reporting.

Reviewer: ```suggestion
The SMART on FHIR `bulk-data-client` open source application provides the core Bulk Data capabilities including kick-off execution, file download, and CLI reporting.
```

`bulk-data-client` is available on [GitHub](https://github.com/smart-on-fhir/bulk-data-client).

### @types/fhir
`bulk-data-export-client` uses the `@types/fhir` package for FHIR type definitions. 

`@types/fhir` is available on [npm](https://www.npmjs.com/package/@types/fhir).

## API Reference

> See [the API Reference](/docs/api-reference)

Reviewer: ```suggestion
See [the API Reference](/docs/api-reference.md)
```

The .md is required for the link to resolve when using the README in GitHub. It still works on the docisfy deployed page.

## Advanced Topics

> See [Advanced Topics](/docs/advanced-topics)

Reviewer: ```suggestion
See [Advanced Topics](/docs/advanced-topics.md)
```

The `.md` is required for the link to resolve when using the README in GitHub. It still works on the docisfy deployed page.

## Documentation Resources

> For more information on the FHIR Bulk Export specifications, refer to the following resources:

Reviewer: ```suggestion
For more information on the FHIR Bulk Export and measure calculation, refer to the following resources:
```
